diff -uN a/storage/src/mozStorageService.cpp b/storage/src/mozStorageService.cpp
index cafa970..93614f0 100644
--- a/storage/src/mozStorageService.cpp
+++ b/storage/src/mozStorageService.cpp
@@ -396,6 +396,7 @@ Service::getSingleton()
     return gService;
   }
 
+#if 0
   // Ensure that we are using the same version of SQLite that we compiled with
   // or newer.  Our configure check ensures we are using a new enough version
   // at compile time.
@@ -411,6 +412,7 @@ Service::getSingleton()
     }
     ::PR_Abort();
   }
+#endif
 
   gService = new Service();
   if (gService) {

--- a/configure.in	2012-07-11 03:32:33 +0800
+++ b/configure.in	2012-07-12 16:18:52 +0800
@@ -129,7 +129,7 @@
 GIO_VERSION=2.18
 STARTUP_NOTIFICATION_VERSION=0.8
 DBUS_VERSION=0.60
-SQLITE_VERSION=3.7.11
+SQLITE_VERSION=3.7.13
 LIBNOTIFY_VERSION=0.4
 
 MSMANIFEST_TOOL=
