--- a/mozilla-build/guess-msvc.bat	2012-07-19 01:44:24 +0800
+++ b/mozilla-build/guess-msvc.bat	2012-11-30 12:28:15 +0800
@@ -141,7 +141,7 @@
 SET SDK61KEY=HKLM\SOFTWARE\Microsoft\Microsoft SDKs\Windows\v6.1
 SET SDK7KEY=HKLM\SOFTWARE\Microsoft\Microsoft SDKs\Windows\v7.0
 SET SDK7AKEY=HKLM\SOFTWARE\Microsoft\Microsoft SDKs\Windows\v7.0A
-SET SDK71KEY=HKLM\SOFTWARE\Microsoft\Microsoft SDKs\Windows\v7.1
+SET SDK71KEY=HKLM\SOFTWARE\Microsoft\Microsoft SDKs\Windows\v7.1A
 SET SDK80KEY=HKLM\SOFTWARE\Microsoft\Windows Kits\Installed Roots
 
 REM Just a base value to compare against
@@ -172,7 +172,7 @@
   IF %ERRORLEVEL% EQU 0 (
     FOR /F "tokens=2*" %%A IN ('REG QUERY "%SDK71KEY%" /v InstallationFolder') DO SET SDKDIR=%%B
     SET SDKVER=7
-    SET SDKMINORVER=1
+    SET SDKMINORVER=1A
   )
 )
 
--- a/mozilla-build/start-msvc11.bat	2012-07-19 01:44:24 +0800
+++ b/mozilla-build/start-msvc11.bat	2012-12-01 08:04:51 +0800
@@ -5,6 +5,8 @@
 SET MOZBUILDDIR=%~dp0
 SET MOZILLABUILD=%MOZBUILDDIR%
 
+SET MOZ_MAXWINSDK=70100
+
 echo "Mozilla tools directory: %MOZBUILDDIR%"
 
 REM Get MSVC paths
@@ -82,7 +84,9 @@
         )
     )
 )
-
+set INCLUDE=%ProgramFiles(x86)%\Microsoft SDKs\Windows\v7.1A\Include;%INCLUDE%
+set PATH=%ProgramFiles(x86)%\Microsoft SDKs\Windows\v7.1A\Bin;%PATH%
+set LIB=%ProgramFiles(x86)%\Microsoft SDKs\Windows\v7.1A\Lib;%LIB%
 cd "%USERPROFILE%"
 
 "%MOZILLABUILD%\msys\bin\bash" --login -i
