--- a/dom/plugins/base/nsPluginHost.cpp	2012-11-21 14:12:02 +0800
+++ b/dom/plugins/base/nsPluginHost.cpp	2012-11-21 16:15:49 +0800
@@ -2798,6 +2798,18 @@
   if (NS_FAILED(rv))
     return rv;
 
+#ifdef TT_MEMUTIL
+  {
+    const char pref_appdir_hash[] = "tete009.appdir_hash";
+    int32_t old_appdir_hash = Preferences::GetInt(pref_appdir_hash);
+    int32_t new_appdir_hash = GetAppDirHash_tt();
+    if (old_appdir_hash != new_appdir_hash &&
+      NS_SUCCEEDED(Preferences::SetInt(pref_appdir_hash, new_appdir_hash))) {
+      pluginReg->Remove(PR_FALSE);
+    }
+  }
+#endif
+
   int64_t fileSize;
   rv = pluginReg->GetFileSize(&fileSize);
   if (NS_FAILED(rv))
