diff -r 3a977cce8359 layout/style/nsCSSParser.cpp
--- a/layout/style/nsCSSParser.cpp	Sat Nov 17 21:27:52 2012 -0500
+++ b/layout/style/nsCSSParser.cpp	Sun Nov 18 13:06:57 2012 +0900
@@ -1072,10 +1072,6 @@
   return NS_OK;
 }
 
-// See Bug 723197
-#ifdef _MSC_VER
-#pragma optimize( "", off )
-#endif
 nsresult
 CSSParserImpl::ParseProperty(const nsCSSProperty aPropID,
                              const nsAString& aPropValue,
@@ -1152,9 +1148,6 @@
   ReleaseScanner();
   return NS_OK;
 }
-#ifdef _MSC_VER
-#pragma optimize( "", on )
-#endif
 
 nsresult
 CSSParserImpl::ParseMediaList(const nsSubstring& aBuffer,
