diff -r 284837bbf631 gfx/thebes/gfxContext.h
--- a/gfx/thebes/gfxContext.h	Sat Aug 25 07:23:12 2012 +0900
+++ b/gfx/thebes/gfxContext.h	Sat Aug 25 10:14:59 2012 +0900
@@ -648,17 +648,21 @@ public:
         /**
          * When this flag is set, snapping to device pixels is disabled.
          * It simply never does anything.
          */
         FLAG_DISABLE_SNAPPING = (1 << 1),
         /**
          * Disable copying of backgrounds in PushGroupAndCopyBackground.
          */
-        FLAG_DISABLE_COPY_BACKGROUND = (1 << 2)
+        FLAG_DISABLE_COPY_BACKGROUND = (1 << 2),
+        /**
+         * tete009 extension
+         */
+        FLAG_DRAW_SINGLE_IMAGE_TT = (1 << 20)
     };
 
     void SetFlag(PRInt32 aFlag) { mFlags |= aFlag; }
     void ClearFlag(PRInt32 aFlag) { mFlags &= ~aFlag; }
     PRInt32 GetFlags() const { return mFlags; }
 
     bool IsCairo() const { return !mDT; }
 
