diff -r 6a3897dad583 mfbt/Attributes.h
--- a/mfbt/Attributes.h	Sat Oct 06 15:30:41 2012 +0900
+++ b/mfbt/Attributes.h	Sat Oct 06 15:38:57 2012 +0900
@@ -8,16 +8,22 @@
 #ifndef mozilla_Attributes_h_
 #define mozilla_Attributes_h_
 
 /*
  * This header does not include any other headers so that it can be included by
  * code that is (only currently) mfbt-incompatible.
  */
 
+#if defined(WIN32) || defined(__SYMBIAN32__)
+#  define TT_RESTRICTED_PTR     __restrict
+#else
+#  define TT_RESTRICTED_PTR     __restrict__
+#endif
+
 /*
  * MOZ_INLINE is a macro which expands to tell the compiler that the method
  * decorated with it should be inlined.  This macro is usable from C and C++
  * code, even though C89 does not support the |inline| keyword.  The compiler
  * may ignore this directive if it chooses.
  */
 #if defined(__cplusplus)
 #  define MOZ_INLINE            inline
